agent/resources/applicationContext.xml

```diff
@@ -17,106 +17,89 @@
-          p:targetObject-ref="agentController"
-          p:targetMethod="loop"/>
+      </bean>
+    </property>
+  </bean>
 
-    <bean id="agentControllerPinger" class="org.springframework.scheduling.timer.MethodInvokingTimerTaskFactoryBean"
-          p:targetObject-ref="agentController"
-          p:targetMethod="ping"/>
+  <bean id="urlService" class="com.thoughtworks.go.util.URLService"/>
+  <bean id="httpService" class="com.thoughtworks.go.util.HttpService"/>
+  <bean id="artifactsManipulator" class="com.thoughtworks.go.publishers.GoArtifactsManipulator"/>
+  <bean id="sslInfrastructureService" class="com.thoughtworks.go.agent.service.SslInfrastructureService"/>
+  <bean id="agentUpgradeService" class="com.thoughtworks.go.agent.service.AgentUpgradeService"/>
+  <bean id="agentRegistry" class="com.thoughtworks.go.config.DefaultAgentRegistry"/>
+  <bean id="subprocessLogger" class="com.thoughtworks.go.util.SubprocessLogger"/>
+  <bean id="agentPluginsInitializer" class="com.thoughtworks.go.agent.AgentPluginsInitializer"/>
+  <bean id="packageAsRepositoryExtension"
+        class="com.thoughtworks.go.plugin.access.packagematerial.PackageAsRepositoryExtension"/>
+  <bean id="scmExtension" class="com.thoughtworks.go.plugin.access.scm.SCMExtension"/>
+  <bean id="taskExtension" class="com.thoughtworks.go.plugin.access.pluggabletask.TaskExtension"/>
+  <bean id="httpClient" class="com.thoughtworks.go.agent.common.ssl.GoAgentServerHttpClient">
+    <constructor-arg ref="httpClientBuilder"/>
+  </bean>
+  <bean id="httpClientBuilder" class="com.thoughtworks.go.agent.common.ssl.GoAgentServerHttpClientBuilder">
+    <constructor-arg ref="systemEnvironment"/>
+  </bean>
 
-    <bean id="agentControllerInstructionExecutor"
-          class="org.springframework.scheduling.timer.MethodInvokingTimerTaskFactoryBean"
-          p:targetObject-ref="agentController"
-          p:targetMethod="executeAgentInstruction"/>
+  <bean id="webSocketClientBuilder" class="com.thoughtworks.go.agent.common.ssl.GoAgentServerWebSocketClientBuilder">
+    <constructor-arg ref="systemEnvironment"/>
+  </bean>
 
+  <bean id="webSocketClientHandler" class="com.thoughtworks.go.agent.WebSocketClientHandler"/>
+  <bean id="webSocketSessionHandler" class="com.thoughtworks.go.agent.WebSocketSessionHandler"/>
 
-    <!-- worker thread -->
-    <bean name="loopThread" class="org.springframework.scheduling.timer.TimerFactoryBean">
-        <property name="scheduledTimerTasks">
-            <list>
-                <bean class="org.springframework.scheduling.timer.ScheduledTimerTask"
-                      p:delay="${agent.get.work.delay}"
-                      p:period="${agent.get.work.interval}"
-                      p:timerTask-ref="agentControllerLooper"/>
-            </list>
-        </property>
-    </bean>
+  <bean id="agentController" class="com.thoughtworks.go.agent.AgentController" factory-method="createInstance">
```

Comment:
Yes I tried this as the first option.
However spring is not able to calculate the dependencies and fails in the object creation order.